src/linear_algebra/tensor_power.lean

```diff
@@ -46,28 +41,240 @@ variables {R : Type*} {M : Type*} [comm_semiring R] [add_comm_monoid M] [module
+begin
+  rw [ghas_mul_def, ghas_one_def],
+  induction a using pi_tensor_product.induction_on with r a x y hx hy,
+  { dsimp only at a,
+    rw [tensor_product.tmul_smul, linear_equiv.map_smul, linear_equiv.map_smul, ←ghas_mul_def,
+      tprod_mul_tprod, cast_tprod],
+    congr' 2 with i,
+    rw fin.elim0'_append,
+    refine congr_arg a (fin.ext _),
+    simp },
+  { rw [tensor_product.tmul_add, map_add, map_add, hx, hy], },
+end
+
+lemma mul_one {n} (a : ⨂[R]^n M) : cast R M (add_zero _) (a ₜ* ₜ1) = a :=
+begin
+  rw [ghas_mul_def, ghas_one_def],
+  induction a using pi_tensor_product.induction_on with r a x y hx hy,
+  { dsimp only at a,
+    rw [←tensor_product.smul_tmul', linear_equiv.map_smul, linear_equiv.map_smul, ←ghas_mul_def,
+      tprod_mul_tprod R a _, cast_tprod],
+    congr' 2 with i,
+    rw fin.append_elim0',
+    refine congr_arg a (fin.ext _),
+    simp },
+  { rw [tensor_product.add_tmul, map_add, map_add, hx, hy], },
+end
+
+lemma mul_assoc {na nb nc} (a : ⨂[R]^na M) (b : ⨂[R]^nb M) (c : ⨂[R]^nc M) :
+  cast R M (add_assoc _ _ _) ((a ₜ* b) ₜ* c) = a ₜ* (b  ₜ* c) :=
+begin
+  let mul : Π (n m : ℕ), (⨂[R]^n M) →ₗ[R] (⨂[R]^m M) →ₗ[R] ⨂[R]^(n + m) M :=
+    (λ n m, (tensor_product.mk R _ _).compr₂ ↑(mul_equiv : _ ≃ₗ[R] ⨂[R]^(n + m) M)),
+  -- replace `a`, `b`, `c` with `tprod R a`, `tprod R b`, `tprod R c`
+  let e : ⨂[R]^(na + nb + nc) M ≃ₗ[R] ⨂[R]^(na + (nb + nc)) M := cast R M (add_assoc _ _ _),
+  let lhs : (⨂[R]^na M) →ₗ[R] (⨂[R]^nb M) →ₗ[R] (⨂[R]^nc M) →ₗ[R] (⨂[R]^(na + (nb + nc)) M) :=
+    (linear_map.llcomp R _ _ _ ((mul _ nc).compr₂ e.to_linear_map)).comp
+      (mul na nb),
+  have lhs_eq : ∀ a b c, lhs a b c = e ((a ₜ* b) ₜ* c) := λ _ _ _, rfl,
+  let rhs : (⨂[R]^na M) →ₗ[R] (⨂[R]^nb M) →ₗ[R] (⨂[R]^nc M) →ₗ[R] (⨂[R]^(na + (nb + nc)) M) :=
+    (linear_map.llcomp R _ _ _ (linear_map.lflip R _ _ _) $
+      (linear_map.llcomp R _ _ _ (mul na _).flip).comp (mul nb nc)).flip,
+  have rhs_eq : ∀ a b c, rhs a b c = (a ₜ* (b ₜ* c)) := λ _ _ _, rfl,
+  suffices : lhs = rhs,
+  from linear_map.congr_fun (linear_map.congr_fun (linear_map.congr_fun this a) b) c,
+  ext a b c,
+  -- clean up
+  simp only [linear_map.comp_multilinear_map_apply, lhs_eq, rhs_eq, tprod_mul_tprod, e,
+    cast_tprod],
+  congr' with j,
+  rw fin.append_assoc,
+  refine congr_arg (fin.append a (fin.append b c)) (fin.ext _),
+  rw [fin.coe_cast, fin.coe_cast],
+end
+
+-- for now we just use the default for the `gnpow` field as it's easier.
+instance gmonoid : graded_monoid.gmonoid (λ i, ⨂[R]^i M) :=
+{ one_mul := λ a, graded_monoid_eq_of_cast (zero_add _) (one_mul _),
+  mul_one := λ a, graded_monoid_eq_of_cast (add_zero _) (mul_one _),
+  mul_assoc := λ a b c, graded_monoid_eq_of_cast (add_assoc _ _ _) (mul_assoc _ _ _),
+  ..tensor_power.ghas_mul,
+  ..tensor_power.ghas_one, }
+
+/-- The canonical map from `R` to `⨂[R]^0 M` corresponding to the algebra_map of the tensor
+algebra. -/
+def algebra_map : R ≃ₗ[R] ⨂[R]^0 M :=
```

Comment:
Should this be `protected`, since I see a `_root_.algebra_map` above?

Reply:
That makes everything quite verbose; I've instead renamed it to `algebra_map₀`.